Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+3. Otherwise throws an error.
+
+Note: `mkAppM` uses `withNewMCtxDepth`, which prevents `Prop` specialization.
+Otherwise the `Subsingleton Prop` instance could apply for example. -/
+def CongrResult.mkDefault (lhs rhs : Expr) : MetaM CongrResult := do
+  if ← isDefEq lhs rhs then
+    return {lhs, rhs, pf? := none}
+  else if let some pf ← (observing? <| mkAppM ``Subsingleton.elim #[lhs, rhs]) then
+    return CongrResult.mk' lhs rhs pf
+  else if let some pf ← (observing? <| mkAppM ``proof_irrel_heq #[lhs, rhs]) then
+    return CongrResult.mk' lhs rhs pf
+  throwError "Could not generated congruence between{indentD lhs}\nand{indentD rhs}"
+
+/-- Does `CongrResult.mkDefault` but makes sure there are no lingering congruence holes. -/
+def CongrResult.mkDefault' (mvarCounterSaved : Nat) (lhs rhs : Expr) : MetaM CongrResult := do
+  if let some h := hasCHole mvarCounterSaved lhs then
+    throwError "Left-hand side{indentD lhs}\nstill has a congruence hole{indentD h}"
+  if let some h := hasCHole mvarCounterSaved rhs then
+    throwError "Right-hand side{indentD rhs}\nstill has a congruence hole{indentD h}"
+  CongrResult.mkDefault lhs rhs
+
+/-- Throw an internal error. -/
+def throwCongrEx (lhs rhs : Expr) (msg : MessageData) : MetaM α := do
+  throwError "congr(...) failed with left-hand side{indentD lhs}\n{""
+    }and right-hand side {indentD rhs}\n{msg}"
+
+/-- If `lhs` or `rhs` is a congruence hole, then process it.
+Only process ones that are at least as new as `mvarCounterSaved`
+since nothing prevents congruence holes from leaking into the local context. -/
+def mkCongrOfCHole? (mvarCounterSaved : Nat) (lhs rhs : Expr) : MetaM (Option CongrResult) := do
+  match cHole? lhs mvarCounterSaved, cHole? rhs mvarCounterSaved with
+  | some (isLhs1, val1, pf1), some (isLhs2, val2, pf2) =>
+    trace[Elab.congr] "mkCongrOfCHole, both holes"
+    unless isLhs1 == true do
+      throwCongrEx lhs rhs "A RHS congruence hole leaked into the LHS"
+    unless isLhs2 == false do
+      throwCongrEx lhs rhs "A LHS congruence hole leaked into the RHS"
+    -- Defeq checks to unify the lhs and rhs congruence holes.
+    unless ← isDefEq (← inferType pf1) (← inferType pf2) do
+      throwCongrEx lhs rhs "Elaborated types of congruence holes are not defeq."
+    if let some (_, lhsVal, _, rhsVal) ← sides? (← inferType pf1) then
+      unless ← isDefEq val1 lhsVal do
+        throwError "Left-hand side of congruence hole is{indentD lhsVal}\n{""
+          }but is expected to be{indentD val1}"
+      unless ← isDefEq val2 rhsVal do
+        throwError "Right-hand side of congruence hole is{indentD rhsVal}\n{""
+          }but is expected to be{indentD val2}"
+    return some <| CongrResult.mk' val1 val2 pf1
+  | some .., none =>
+    throwCongrEx lhs rhs "Right-hand side lost its congruence hole annotation."
+  | none, some .. =>
+    throwCongrEx lhs rhs "Left-hand side lost its congruence hole annotation."
```

Comment:
Similarly here.

Reply:
This is also not impossible, sadly. We always have the problem that the congruence hole could be given to an elaborator that reduces its argument in some way, stripping the annotation. These particular cases come up when that elaborator does it conditionally (what if `h : 1 = x` and `foo% 1` turns `1` into `Nat.succ Nat.zero` but leaves `foo% x` alone since it's not a number?).